### PR TITLE
Implementing Course Component

### PR DESCRIPTION
**Pull request by MuhammedAdelTaha**
- Just to test If there are any conflicts.
- I rearranged the hierarchy of entities in the entity folder to be clearer. 
- Please notice the changes and tell me if there are any mistakes. 
- I am still working to complete **CoursePrerequisiteService** and **StudentCourseGradeService** and the **Tests**. 
- I will leave the rest for you guys.

### CODE REVIEW

File: Springboot/src/main/java/com/centralstudenthub/error/CourseAlreadyExistsException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.centralstudenthub.error;
+
+public class CourseAlreadyExistsException extends Exception{
```
Comment (Ali-Mones) [inline Springboot/src/main/java/com/centralstudenthub/error/CourseAlreadyExistsException.java:3]: add a space before the {

File: Springboot/src/main/java/com/centralstudenthub/error/CourseNotFoundException.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.centralstudenthub.error;
+
+public class CourseNotFoundException extends Exception{
```
Comment (Ali-Mones) [inline Springboot/src/main/java/com/centralstudenthub/error/CourseNotFoundException.java:3]: Add space before the {

File: Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.centralstudenthub.error;
+
+public class ConflictException extends Error {
```
Comment (Ali-Mones) [inline Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java:3]: extends Exception not Error

File: Springboot/src/main/java/com/centralstudenthub/error/AllCoursesAlreadyExistsException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.centralstudenthub.error;
+
+public class AllCoursesAlreadyExistsException extends Exception{
```
Comment (Ali-Mones) [inline Springboot/src/main/java/com/centralstudenthub/error/AllCoursesAlreadyExistsException.java:3]: Add space before the {

File: Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+    public ConflictException(String message) {
```
Comment (Ali-Mones) [inline Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java:4]: This class name is too vague
What does Conflict mean ?

**Review by Ali-Mones — CHANGES_REQUESTED** [review]

**Review by Ali-Mones — CHANGES_REQUESTED** [review]
Please change the name of the "error" package into "exception"

File: Springboot/src/main/java/com/centralstudenthub/error/CourseAlreadyExistsException.java (hunk 1 above)
Comment (MuhammedAdelTaha) [reply Springboot/src/main/java/com/centralstudenthub/error/CourseAlreadyExistsException.java]: Done.

File: Springboot/src/main/java/com/centralstudenthub/error/CourseNotFoundException.java (hunk 2 above)
Comment (MuhammedAdelTaha) [reply Springboot/src/main/java/com/centralstudenthub/error/CourseNotFoundException.java]: Done.

File: Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java (hunk 3 above)
Comment (MuhammedAdelTaha) [reply Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java]: Done.

File: Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java (hunk 5 above)
Comment (MuhammedAdelTaha) [reply Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java]: I implemented three types of exceptions.

- **Not Found Exception**: if I am trying to get record not in the database.
- **Already Exists Exception**: if I am trying to put record already in the database.
- **Conflict Exception**: if I am trying to put a record can't be in the database. Like assigning a course as a prerequisite for itself.

**Comment by MuhammedAdelTaha** [comment]
> Please change the name of the "error" package into "exception"

Done.

File: Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java (hunk 5 above)
Comment (Ali-Mones) [reply Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java]: Maybe change name to DatabaseLogicalConstraintException ?
The name Conflict can carry many meanings

Comment (MuhammedAdelTaha) [reply Springboot/src/main/java/com/centralstudenthub/error/ConflictException.java]: Good suggestion ✌️

**Review by Ali-Mones — APPROVED** [review]

**Review by MarwanElsays — APPROVED** [review]
good job

**Pull request merged**